src/main/java/com/angelbroking/smartapi/http/SmartAPIResponseHandler.java

```diff
@@ -87,10 +99,64 @@ private SmartAPIException dealWithException(JSONObject jsonObject, String code)
 		case "AB1001":
 		case "AB1011":
 			return new PermissionException(jsonObject.getString("message"), code);
-
+            case "AG8004":
+                return new ApiKeyException(APIKEY_EXCEPTION_MESSAGE, code);
 		default:
 			return new SmartAPIException(jsonObject.getString("data not found"));
 		}
 	}
 
+    public String handler(Response response, String body) throws SmartAPIException, JSONException, IOException {
+        if (response.code() == 200) {
+            return handleResponse(response,body);
+        } else if (response.code() == 400){

```

Comment:
This condition is wroing. If response code is 400 then does not mean Invalid exchange provided. Please provide a valid exchange.

400 mean bad input and any request tag may not have value.

Reply:
The error message for the 400 status code is mentioned in SSS-41,

Reply:
have updated the error message to 

Bad request. Please provide valid input